framework/src/tinyalsa/tinyalsa.c

```diff
@@ -532,50 +533,82 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
 #else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
 #endif
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
-		return -errno;
-	} else {
-		pcm->running = 1;
+	if (!pcm->running) {
+		pcm->enqueuedBytes = 0;
+		/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		bufdesc.session = pcm->session;
+#endif
+#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
+		bufdesc.numbytes = pcm->config.period_size;
+		for (pcm->bufPtr = 0; pcm->bufPtr < buf_info.nbuffers; pcm->bufPtr++)
+#else
+		bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
+		for (pcm->bufPtr = 0; pcm->bufPtr < CONFIG_AUDIO_NUM_BUFFERS; pcm->bufPtr++)
```

Comment:
We must not initialize bufPtr to 0 here. According to your logic, we need to retain the bufPtr values during consecutive calls to readi.

Reply:
Actually, here I have copied the code from pcm_start .
It will be when the first pcm_readi is called. Then pcm_start would set pcm->running flag to true. Then for successive calls to pcm_readi , the enqueue buffer shouldnt run in this area.

Reply:
That means, if user requests small amount of data, say 100 bytes on the first call to readi, then you will use only 1 of the 4 buffers. Next time onwards, even if user requests large amount of data, say 8000 bytes, you cannot use any of the remaining buffers.

Reply:
My latest commit addresses this